ydb/core/persqueue/partition.cpp

```diff
@@ -1624,113 +1618,168 @@ size_t TPartition::GetUserActCount(const TString& consumer) const
-            KVWriteInProgress = true;
-            HaveWriteMsg = true;
-        } else {
-            AnswerCurrentWrites(ctx);
-            AnswerCurrentReplies(ctx);
-        }
-
+    if (WriteCycleSizeEstimate >= MAX_WRITE_CYCLE_SIZE || WriteKeysSizeEstimate >= MAX_USERS) {
+        SwitchToCommitState();
         return;
     }
-
-    auto visitor = [this, &ctx, &request](auto& event) {
-        using T = std::decay_t<decltype(event)>;
-        if constexpr (TIsSimpleSharedPtr<T>::value) {
-            return this->ProcessUserActionOrTransaction(*event, request.Get(), ctx);
-        } else {
-            return this->ProcessUserActionOrTransaction(event, request.Get(), ctx);
-        }
+    auto visitor = [this](auto& event) {
+        return this->PreProcessUserActionOrTransaction(event);
     };
 
     FirstEvent = true;
     for (bool stop = false; !stop && !UserActionAndTransactionEvents.empty(); ) {
         auto& front = UserActionAndTransactionEvents.front();
 
-        switch (std::visit(visitor, front)) {
-        case EProcessResult::Continue:
-        case EProcessResult::Reply:
+        switch (std::visit(visitor, front.Event)) {
+        case EProcessResult::ContinueAutoCommit:
+            front.State = ECommitState::Committed;
+            [[fallthrough]];
+        case EProcessResult::ContinueWaitCommit:
+            MoveUserActOrTxToCommitState();
+            break;
+        case EProcessResult::ContinueDrop:
             UserActionAndTransactionEvents.pop_front();
-            FirstEvent = false;
             break;
         case EProcessResult::Break:
+            MoveUserActOrTxToCommitState();
             stop = true;
             break;
-        case EProcessResult::Abort:
+        case EProcessResult::Blocked:
+            stop = true;
+            break;
+        case EProcessResult::NotReady:
             return;
         }
+        FirstEvent = false;
+        TxOrUserActProcessed = true;
+    }
+    SwitchToCommitState();
+}
+
+void TPartition::MoveUserActOrTxToCommitState() {
+    auto& front = UserActionAndTransactionEvents.front();
+    UserActionAndTxPendingCommit.push_back(std::move(front));
+    UserActionAndTransactionEvents.pop_front();
+}
+
+void TPartition::ProcessCommitQueue() {
+    Y_ABORT_UNLESS(WaitingForCommit);
+    Y_ABORT_UNLESS(!KVWriteInProgress);
+    if (!PersistRequest) {
+        PersistRequest = MakeHolder<TEvKeyValue::TEvRequest>();
     }
 
+    bool commit = true;
+    auto visitor = [this, request=PersistRequest.Get(), &commit](auto& event) {
+        return this->ExecUserActionOrTransaction(event, request, commit);
+    };
+    while (!UserActionAndTxPendingCommit.empty()) {
+        auto& front = UserActionAndTxPendingCommit.front();
+        auto& event = front.Event;
+        switch (front.State) {
+            case ECommitState::Pending:
+                return;
+            case ECommitState::Aborted:
+                commit = false;
```

Comment:
Why is commit global variable, not per UserActionAndTxPendingCommit?

Reply:
This is the state of single UserAction.